### PR TITLE
Shifter

### PR DESCRIPTION
**Pull request by Wunarg**
The shifter is a new building in the automation section, it does not have collision and is 1x1 big, when an entity collides with it that entity is moved in the direction of this rotatable building. After this the shifter starts a 120 tick cooldown. 
The force of the shifter is adjusted for mass and decreases to a minnimum the higher the velocity of the entity already is.
-Costs 20 wood, 10 stone and 2 copper coils (you usually need many of these)
Notes:
-the cooldown ensures that this does not protect against bombs greatly (one may be deflected or slowed to a a soft landing)
-the sprite changes slightly when recharge to indicate this
-the force is not strong enough to stop people from walking down a passage filled with them
-roughly 5 layers are needed to slow down a bombs from a low ish height
-particle accelerators are technicly possible (aka shifter circles) but extremely difficult due to corners and the cooldown
--Very fun
-Allows for many many fun things like elevators, people launchers, trampolines, railguns etc

### CODE REVIEW

File: TFlippy_TerritoryControl_Characters_Dev/Entities/Characters/Builder/CommonBuilderBlocks.as [hunk 1]
```diff
@@ -587,6 +587,13 @@ void addCommonBuilderBlocks(BuildBlock[][]@ blocks, int teamnum = 7)
 		b.size.Set(24, 24);
 		blocks[2].push_back(b);
 	}
+		{
```
Comment (Gingerbeard5773) [inline TFlippy_TerritoryControl_Characters_Dev/Entities/Characters/Builder/CommonBuilderBlocks.as]: Indent oopsie

**Review by Gingerbeard5773 — COMMENTED** [review]
Shifter is an interesting concept, however I believe that it is unfit for the mod because of its exploitable nature. Everything that the shifter does can be done by other parts of the mod and so I believe it is unneeded.

File: TFlippy_TerritoryControl_Structures_Dev/Entities/Structures/Factories/Shifter/Shifter.as [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	this.Tag("builder always hit");
+
+	this.set_u32("charge",0);
+	CSprite@ sprite = this.getSprite();
+	sprite.SetAnimation("off"); //Starts offline
+	sprite.SetZ(-100.0f);
+}
+
+void onTick(CBlob@ this)
```
Comment (Vam-Jam) [inline TFlippy_TerritoryControl_Structures_Dev/Entities/Structures/Factories/Shifter/Shifter.as]: Optimization: Remove onTick and move the logic into onCollosion

Store last time used and if it's more > 120 ticks then push the blob


Put sprite animation into onTick(CSprite@)  so only clients will run that code

File: TFlippy_TerritoryControl_Structures_Dev/Entities/Structures/Factories/Shifter/Shifter.as [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	{
+		CSprite@ sprite = this.getSprite();
+		sprite.SetAnimation("on");
+		sprite.SetZ(-100.0f);
+	}
+}
+
+void onCollision(CBlob@ this, CBlob@ blob, bool solid)
+{
+	if (this.get_u32("charge")>=120)
```
Comment (Vam-Jam) [inline TFlippy_TerritoryControl_Structures_Dev/Entities/Structures/Factories/Shifter/Shifter.as]: charge") space >= space 120

**Review by Vam-Jam — CHANGES_REQUESTED** [review]

**Comment by Wunarg** [comment]
Do not add this one yet on my end it has crashed sometimes randomly for a completely unknown reason and while it might be related to my computer specificly i would like to figure it out before unleashing it on the server.

**Comment by laikadoggirl** [comment]
shifter added when

**Comment by laikadoggirl** [comment]
just ignore dragon and ginger they just hate anything "fun" because it wont let them tryhard and kill neutrals harder

**Comment by laikadoggirl** [comment]
vamist you can oh whats this ban button do all you want, but you know for a fact something not having a use aside from being fun is a retarded reason not to add it

**Comment by laikadoggirl** [comment]
Heres a novel idea, what if we added fun and engaging things to tc, instead of just shitting about the faction neutral balance for all eternity

**Pull request merged**